src/colr.cc

```diff
@@ -0,0 +1,1069 @@
+  COMPOSITE_HSL_LUMINOSITY = 27,  // https://www.w3.org/TR/compositing-1/#blendingluminosity
+};
+
+bool ParseAffine(const ots::Font* font,
+                 const uint8_t* data, size_t length,
+                 colrState& state, bool var)
+{
+  auto& set = var ? state.varAffines : state.affines;
+  if (setContains(set, data)) {
+    return true;
+  }
+  set.insert(data);
+
+  ots::Buffer subtable(data, length);
+
+  Fixed xx, yx, xy, yy, dx, dy;
+  VarIdxBase varIndexBase;
+
+  if (!subtable.ReadS32(&xx) ||
+      !subtable.ReadS32(&yx) ||
+      !subtable.ReadS32(&xy) ||
+      !subtable.ReadS32(&yy) ||
+      !subtable.ReadS32(&dx) ||
+      !subtable.ReadS32(&dy) ||
+      (var && !subtable.ReadU32(&varIndexBase))) {
+    return OTS_FAILURE_MSG("Failed to read [Var]Affine2x3");
+  }
+
+  return true;
+}
+
+// Paint-record dispatch function that reads the format byte and then dispatches
+// to one of the record-specific helpers.
+bool ParsePaint(const ots::Font* font, const uint8_t* data, size_t length, colrState& state);
+
+// All these paint record parsers start with Skip(1) to ignore the format field,
+// which the caller has already read in order to dispatch here.
+
+bool ParsePaintColrLayers(const ots::Font* font,
+                          const uint8_t* data, size_t length,
+                          colrState& state)
+{
+  ots::Buffer subtable(data, length);
+
+  uint8_t numLayers;
+  uint32_t firstLayerIndex;
+
+  if (!subtable.Skip(1) ||
+      !subtable.ReadU8(&numLayers) ||
+      !subtable.ReadU32(&firstLayerIndex)) {
+    return OTS_FAILURE_MSG("Failed to read PaintColrLayers record");
+  }
+
+  if (uint64_t(firstLayerIndex) + numLayers > state.numLayers) {
+    return OTS_FAILURE_MSG("PaintColrLayers exceeds bounds of layer list");
+  }
+
+  return true;
+}
+
+bool ParsePaintSolid(const ots::Font* font,
+                     const uint8_t* data, size_t length,
+                     colrState& state, bool var)
+{
+  ots::Buffer subtable(data, length);
+
+  uint16_t paletteIndex;
+  F2DOT14 alpha;
+
+  if (!subtable.Skip(1) ||
+      !subtable.ReadU16(&paletteIndex) ||
+      !subtable.ReadS16(&alpha)) {
+    return OTS_FAILURE_MSG("Failed to read PaintSolid");
+  }
+
+  if (paletteIndex >= state.numPaletteEntries && paletteIndex != 0xffffu) {
+    return OTS_FAILURE_MSG("Invalid palette index %u PaintSolid", paletteIndex);
+  }
+
+  if (alpha < 0 || alpha > F2DOT14_one) {
+    OTS_WARNING("Alpha value outside valid range 0.0 - 1.0");
+  }
+
+  if (var) {
+    VarIdxBase varIndexBase;
+    if (!subtable.ReadU32(&varIndexBase)) {
+      return OTS_FAILURE_MSG("Failed to read PaintVarSolid");
+    }
+  }
+
+  return true;
+}
+
+bool ParsePaintLinearGradient(const ots::Font* font,
+                              const uint8_t* data, size_t length,
+                              colrState& state, bool var)
+{
+  ots::Buffer subtable(data, length);
+
+  uint32_t colorLine;
```

Comment:
This should be `colorLineOffset` to make it clear it is an offset.

Reply:
Done.